rclcpp/include/rclcpp/experimental/intra_process_manager.hpp

```diff
@@ -352,17 +362,46 @@ class IntraProcessManager
       auto subscription_base = subscription_it->second.lock();
       if (subscription_base) {
         auto subscription = std::dynamic_pointer_cast<
-          rclcpp::experimental::SubscriptionIntraProcessBuffer<MessageT, Alloc, Deleter>
+          rclcpp::experimental::SubscriptionIntraProcessBuffer<PublishedType,
+          PublishedTypeAllocator, PublishedTypeDeleter, ROSMessageType>
           >(subscription_base);
         if (nullptr == subscription) {
```

Comment:
this code is very complex to read due the abundance of nested conditions.
I recommend to re-organize it to reduce nesting, for example by inverting many checks and completely resolving some code paths as soon as possible.

For example:
```
if (!subscription_base) {
  subscriptions_.erase(id);
  continue;
}

auto subscription = std::dynamic...
if (subscription) {
  subscription->provide_intra_process_data(message);
  continue;
}

auto ros_message_subscription = std::dynamic...

```

Reply:
https://github.com/ros2/rclcpp/pull/1849/commits/1779cd820a78763c867ad8183166c7025e5285fa refactors this in the way you suggested, by getting out early whenever we can.  I'm going to mark this conversation as resolved, since it is much easier to read now.  But feel free to comment again and unresolve the conversation if you disagree.